docs/proposals/004-com-support.md

```diff
@@ -0,0 +1,240 @@
+
+While COM provides idioms for addressing many practical challenges, it is also inconvenient in that it introduces a lot of *boilerplate*:
+
+* COM types all need to implement the core `IUnknown` operations for casting/querying and reference counting.
+
+* Code using COM interfaces needs to perform `AddRef` and `Release` operations manually, or use smart pointer types to automate lifetime management.
+
+* Code that calls into COM interfaces typically needs to use boilerplate code and/or macros to deal with `HRESULT` error codes, handling or propagating them as needed.
+
+Our in-progress work on supporting CPU programming in Slang emphasizes supporting idiomatic code without a lot of boilerplate.
+Our intended path includes things that are compatible with the COM philosophy, like reference-counted lifetime management and idiomatic use of result/error codes, but those features don't currently align with the more explicit style used by COM in C/C++.
+
+Related Work
+------------
+
+The .NET platform includes some support for allowing .NET `interface`s and COM interfaces to interoperate.
+TODO: Need to study this and learn how it works.
+
+Proposed Approach
+-----------------
+
+We propose to allow COM interfaces to be declared using the Slang `interface` construct, with an appropriate attribute or modifier:
+
+```
+[COM] interface IDevice
+{
+    ITexture createTexture(__read TextureDesc desc) throws HRESULT;
+
+    void setTexture(int index, ITexture texture);
+}
+```
+
+A declaration like the above will translate into output C++ along the lines of:
+
+```
+struct IDevice : public IUnknown
+{
+    virtual HRESULT SLANG_MCALL createTexture(TextureDesc const& desc, ITexture** _result) = 0;
+
+    virtual void SLANG_MCALL setTexture(int index, ITexture* texture) = 0;
+};
+```
+
+Key things to note:
+
+* The `[COM] interface` becomes a C++ `struct` that inherits from `IUnknown`
+* Methods defined in the `[COM] interface` become pure-virtual `SLANG_MCALL` methods in C++
+* Parametes/values of a `[COM] interface` type `IFoo` in Slang translate to `IFoo*` in C++
+* Methods that have a `throws HRESULT` clause are transformed to have an `HRESULT` return type and an output parameter for their result
+
+A Slang `class` can declare that it implements zero or more `[COM] interface`s. Code like this:
+
+```
+class MyTexture : ITexture
+{
+    // ...
+}
+
+class MyDevice : IDevice
+{
+    ITexture createTexture(__read TextureDesc desc) throws HRESULT
+    {
+        return ...;
+    }
+
+    void setTexture(int index, ITexture texture)
+    {
+        ...;
+    }
+}
+```
+
+translates into output C++ like this:
+
+```
+class MyTexture : public slang::Object, public ITexture
+{
+    // ...
+};
+
+class MyDevice : public slang::Object, public  IDevice
+{
+    virtual HRESULT QueryInterface(REFIID riid, void **ppvObject) { ... }
+    virtual ULONG AddRef() { ... }
```

Comment:
In IDL, there is markup to say `ppvObject` (aka `outObject` ;0) is of the type specified in riid.

Reply:
I suspect that we should have a plan within this `[COM] interface` stuff for how the `REFIID riid, void** ppvObject` idiom can be expressed in terms of generics. *If* we were to define `QueryInterface` at the Slang level, I would like it if it could look something like:

```
T QueryInterface< interface T : IUnknown >() throws HRESULT;
```

The basic idea being that `QueryInterface` either signals a failure through the `HRESULT` or it returns a valid instance of `T`. The pitch here would be that `<interface T>` would be a way to define a generic parameter that only accepted `interface` types as arguments (and the `IUnknown` constraint ensures that `T` is a sub-interface of `IUnknown` and thus a `[COM]` interface).

There are certainly various pieces of the D3D and DXGI APIs that rely on the `REFIID, void**` idiom, so being able to translate those over to something idiomatic in Slang would be great.